test/hotspot/jtreg/compiler/c2/irTests/UDivINodeIdealizationTests.java

```diff
@@ -0,0 +1,148 @@
+        assertResult(min, min, false);
+        assertResult(max, max, false);
+    }
+
+    @DontCompile
+    public void assertResult(int a, int b, boolean shouldThrow) {
+        try {
+            Asserts.assertEQ(Integer.divideUnsigned(a, a), constant(a));
+            Asserts.assertFalse(shouldThrow, "Expected an exception to be thrown.");
+        } catch (ArithmeticException e) {
+            Asserts.assertTrue(shouldThrow, "Did not expect an exception to be thrown.");
+        }
+
+        try {
+            Asserts.assertEQ(Integer.divideUnsigned(a, Integer.divideUnsigned(b, b)), identityThird(a, b));
+            Asserts.assertFalse(shouldThrow, "Expected an exception to be thrown.");
+        } catch (ArithmeticException e) {
+            Asserts.assertTrue(shouldThrow, "Did not expect an exception to be thrown.");
+        }
+
+        try {
+            Asserts.assertEQ(Integer.divideUnsigned((a * b), b), retainDenominator(a, b));
+            Asserts.assertFalse(shouldThrow, "Expected an exception to be thrown.");
+        } catch (ArithmeticException e) {
+            Asserts.assertTrue(shouldThrow, "Did not expect an exception to be thrown.");
+        }
+
+        Asserts.assertEQ(Integer.divideUnsigned(a, 1), identity(a));
+        Asserts.assertEQ(Integer.divideUnsigned(a, Integer.divideUnsigned(13, 13)), identityAgain(a));
+        Asserts.assertEQ(Integer.divideUnsigned(a, 8), divByPow2(a), "divByPow2 " + a);
+        Asserts.assertEQ(Integer.divideUnsigned(a, -2147483648), divByPow2Big(a));
+        Asserts.assertEQ(Integer.divideUnsigned(a, Integer.divideUnsigned(-2129457054, -2129457054)), identityAgainButBig(a));
+    }
+
+    @Test
+    @IR(failOn = {IRNode.DIV})
+    @IR(counts = {IRNode.DIV_BY_ZERO_TRAP, "1"})
+    // Checks x / x => 1
+    public int constant(int x) {
+        return Integer.divideUnsigned(x, x);
+    }
+
+    @Test
+    @IR(failOn = {IRNode.DIV})
+    // Checks x / 1 => x
+    public int identity(int x) {
+        return Integer.divideUnsigned(x, 1);
+    }
+
+    @Test
+    @IR(failOn = {IRNode.DIV})
+    // Checks x / (c / c) => x
+    public int identityAgain(int x) {
+        return Integer.divideUnsigned(x, Integer.divideUnsigned(13, 13));
+    }
+
+    @Test
+    @IR(failOn = {IRNode.DIV})
+    // Checks x / (c / c) => x
+    public int identityAgainButBig(int x) {
+        return Integer.divideUnsigned(x, Integer.divideUnsigned(-2129457054, -2129457054));  // Integer.parseUnsignedInt("2165510242") = -2129457054
+    }
+
+    @Test
+    @IR(failOn = {IRNode.DIV})
+    @IR(counts = {IRNode.DIV_BY_ZERO_TRAP, "1"})
+    // Checks x / (y / y) => x
+    public int identityThird(int x, int y) {
+        return Integer.divideUnsigned(x, Integer.divideUnsigned(y, y));
+    }
+
+    @Test
+    @IR(counts = {IRNode.MUL, "1",
+            IRNode.UDIV_I, "1",
+            IRNode.DIV_BY_ZERO_TRAP, "1"
+    })
+    // Hotspot should keep the division because it may cause a division by zero trap
```

Comment:
This comment is actually incorrect, the division is kept because the transformation is wrong.

Reply:
Which part of the transformation do you think is wrong? Do you think there might also be overflow issues and that's why it cannot be replaced with x?

(I took over this comment and test from the pre-existing test for the signed version: test/hotspot/jtreg/compiler/c2/irTests/DivINodeIdealizationTests.java.)

Reply:
Yupp it can overflow, if it does not then we can totally do this transformation. I think you should remove this line from the newly added test, feel free to also remove it from the old one.